### PR TITLE
✨ - Add Mono/Unity support and extend constraints support

### LINKED ISSUE #10: Support for Unity
Please forgive me if this isn't the place to ask this question. I am trying to use this within Unity and my script fails to locate libnlopt-0.dll. Will this work as is within Unity?

### PR DESCRIPTION
**Pull request by JeremyBois**
- Add mono and Unity3d support
  - Add support for both
    - nlopt_add_equality_mconstraint 
    - nlopt_add_inequality_mconstraint 

Closes #10

### CODE REVIEW

File: NLoptNet/NLoptSolver.cs [hunk 1]
```diff
@@ -9,12 +14,47 @@ namespace NLoptNet
...
-		private delegate double nlopt_func(uint n, [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 0), In] double[] x, [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 0), In, Out] double[] gradient, IntPtr data);
-		[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
-		private delegate void nlopt_mfunc(uint m, [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 0), In, Out] double[] result, uint n, [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2), In] double[] x, [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2), In, Out] double[] gradient, IntPtr data);
-
-		[DllImport("nlopt", CallingConvention = CallingConvention.Cdecl)]
+#if MONO
+                [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+                private delegate double nlopt_func(
+                    uint n,
+                    IntPtr p_x,
```
Comment (BrannonKing) [inline NLoptNet/NLoptSolver.cs:21]: Why not use the MONO approach for all frameworks? Is there any performance difference with the MarshalAs approach?

File: NLoptNet/NLoptSolver.cs [hunk 2]
```diff
@@ -9,12 +14,47 @@ namespace NLoptNet
...
+            uint m,
+            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 0), In, Out] double[] result,
+            uint n,
+            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2), In] double[] x,
+            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2), In, Out] double[] gradient,
+            IntPtr data
+        );
+#endif
+
+        [DllImport("nlopt", CallingConvention = CallingConvention.Cdecl)]
```
Comment (BrannonKing) [inline NLoptNet/NLoptSolver.cs:57]: Use tabs.

**Review by BrannonKing — CHANGES_REQUESTED** [review]
Do we need to upgrade the deployment frameworks to include Mono somehow?

File: NLoptNet/NLoptSolver.cs (hunk 1 above)
Comment (JeremyBois) [reply NLoptNet/NLoptSolver.cs]: I suppose the .Net approach with `In` and `Out` is only there to remove boilerplate code required with mono (not an expert). However currently the mono approach required allocation of a new bloc of memory each time a wrapped function is called. Maybe the .Net way performs better memory management but again not an expert. 

I would tend to say that is better to keep the .Net approach when possible for the explained reasons.

**Comment by JeremyBois** [comment]
> 
> 
> Do we need to upgrade the deployment frameworks to include Mono somehow?

I am not sure to understand what you means. If you want to be able to run tests with mono I cannot help because my understanding of Visual Studio is very limited (not using it at all ...). There is an installer for mono on Windows [here](https://www.mono-project.com/download/stable/)

**Pull request merged**